### ISSUE TITLE
Hexadecimal literals are broken

### ISSUE THREAD

**Issue by mode777** [wren-lang/wren#926]
I noticed hex literals no longer work. Consider this code
```
var SDL_OPENGL = 0x00000002
System.print(SDL_OPENGL)
``` 
On 0fa16a20ecfab7108ab047cb64c7085fb7492f0a
> 2

On 3a06580b89459553b866f27eb7ee8792b0310a76
> Error: This code should not be reached in wrenGetClassInline()

Finally, on  9dfbc021a03d6313c48d8975f5168db740ce7e0d
> 0

This was tested on mingw (32bit) on windows 10

**Comment by mhermier** [comment]
Seems to be specific to mingw: http://mingw.5.n7.nabble.com/strtod-is-not-C99-compatible-td14162.html

**Comment by mode777** [comment]
I can confirm that.
This program outputs 0.000
```
/* strtod example */
#include <stdio.h>      /* printf, NULL */
#include <stdlib.h>     /* strtod */

int main ()
{
  char input[] = "0x02";
  double d1 = strtod (input, NULL);
  printf ("%f\n", d1);
  return 0;
}
```

So what doeas that mean for Wren? No MinGW support?

**Comment by ruby0x1** [comment]
No, it means we shouldn't use unportable code. I've reverted the commit that introduced it for now. Thanks for testing.

**Comment by mhermier** [comment]
According to our documentation C99 is required, so strictly speaking mingw should be unsupported.
Can you tell me what the value of the __STDC_VERSION__ in that environment to see if we can mitigate the issue with that ?

**Comment by mhermier** [comment]
We can mitigate by doing something like this. I think it should be less troublesome on conformant platforms.

```
From c3b81d7e6bf6d8fe0a5361b663d083d736376057 Mon Sep 17 00:00:00 2001
From: Michel Hermier <michel.hermier@gmail.com>
Date: Sat, 13 Feb 2021 21:00:40 +0100
Subject: [PATCH 1/1] wren/compiler: On C99 `strtod` also parses hex values.
 Ignore non conformant implementation (mingw).

---
 src/vm/wren_compiler.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/vm/wren_compiler.c b/src/vm/wren_compiler.c
index 95c470dd..6aea0047 100644
--- a/src/vm/wren_compiler.c
+++ b/src/vm/wren_compiler.c
@@ -717,11 +717,13 @@ static void makeNumber(Parser* parser, bool isHex)
 {
   errno = 0;
 
+#if __STDC_VERSION__ < 199901L || defined(__MINGW32__) || defined(__MINGW64__)
   if (isHex)
   {
     parser->next.value = NUM_VAL((double)strtoll(parser->tokenStart, NULL, 16));
   }
   else
+#endif
   {
     parser->next.value = NUM_VAL(strtod(parser->tokenStart, NULL));
   }
-- 
2.30.1

```

**Comment by mode777** [comment]
__STDC_VERSION__ is defined as 201710 which should be C17.

@mhermier I tried to apply your patch but it failed i'm not very familiar with applying git patches

**Comment by mhermier** [comment]
Should apply with `patch -p1` otherwise I think it should be `git am` to
insert it in your local branch.
Odd that this issue is still not addressed in mingw, while claiming C17...

**Comment by mode777** [comment]
The patch worked, thanks! 
Yeah, there's a lot odd about mingw. It seems abandoned. Event the domain is expired. I wonder if this issue persists in mingw64 as well. I don't have it installed though.

**Comment by mhermier** [comment]
It seems mingw32 seems pretty abandoned since 2015, and last commit seems to be from 2018... If you can migrate to mingw64 to test even without the revert so we can call it a day and drop support for mingw32, it would be nice. But there is no obligation.

**Comment by PureFox48** [comment]
It appears that MinGW (mingw32 as was) is still being developed but is hosted nowadays at [OSDN](https://osdn.net/projects/mingw/) rather than SourceForge.

**Comment by mhermier** [comment]
We will have to bring a function to replace strtod, even without the messy support of mingw and probably other platform/support/etc, its behavior is dependent of the locale definition, which is a no go for the compiler.

**Comment by mhermier** [comment]
@mode777 could you check if using `parser->next.value = NUM_VAL(strtod_l(parser->tokenStart, NULL), newlocale(LC_ALL_MASK, "C", NULL));` is conformant on mingw ? Not a very viable solution since it relies on POSIX-2008, but can be a partial solution.

**Comment by mode777** [comment]
@mhermier Sorry I didn't comearound to test this earlier. Doesn't work, seems like Windows is not POSIX compatible ;)

**Comment by chm8d** [comment]
Seems like number parsing is implemented in [src/vm/wren_compiler.c](https://github.com/wren-lang/wren/blob/main/src/vm/wren_compiler.c).

Wouldn't it be better to use the result from `readHexDigit` to build the number directly in `readHexNumber`? We could have `readHexNumber` build the integer, then pass it to `makeNumber` as an argument rather than parsing it using `strtod`.

```c
static void readHexNumber(Parser* parser)
{
  // Skip past the `x` used to denote a hexadecimal literal.
  nextChar(parser);

  int number = 0;
  int digit = -1;

  // Build the number from the digits we read
  while ((digit = readHexDigit(parser)) != -1) {
    number = number * 16 + digit;
  }

  makeNumber(parser, (double)number);
}
```

`makeNumber` would need to be rewritten, something like this:

```c
static void makeNumber(Parser* parser, double number)
{
  parser->next.value = NUM_VAL(number);
  makeToken(parser, TOKEN_NUMBER);
}
```

Of course, the function `readNumber` would also need a rewrite. I think I could rewrite them and do a pull request if you want.

**Comment by mhermier** [comment]
While technically valid, maybe instead it should be better to use the `endptr` output parameter to adjust `parser->currentChar` and bypass the double scanning. It would also bring allow to parse parsing of hex exponents for free (though it is mostly usefull for NaN and we forbidd qNaN usage...).

**Comment by fermian** [comment]
> We will have to bring a function to replace strtod, even without the messy support of mingw and probably other platform/support/etc, its behavior is dependent of the locale definition, which is a no go for the compiler.

Parsing floating-point numbers directly isn't a trivial task, so I think we should port a strtod implementation from another library. [Python](https://github.com/python/cpython/blob/39f643614d03748a5fad462fe7ed26a174a522fa/Python/dtoa.c) and [Ruby](https://github.com/ruby/ruby/blob/4b6347ab3907850769aa8e93674a4c0a4a18c668/missing/dtoa.c) ported strtod from [the netlib library](http://www.netlib.org/fp/dtoa.c), should Wren do the same?

**Comment by mhermier** [comment]
It is an idea, but the problem is not trivial. Parsing a token that it is believed to be a number, and parsing a string to a number, while being a very coupled problem are 2 different problems.

eg `0000049` can be considered a valid number parsing wise so that the error can happens when interpreting it.
